STIPS/STIP-005.md

```diff
@@ -0,0 +1,1983 @@
+    + openNotionalFraction = (positionInfo.quoteBalance  * -1 ) + reducedOpenNotional 
+
+  + Account for already accrued PnL from non-issuance/redemption sources (ex: levering, liquidation) 
+
+    + totalFundingAndCarriedPnL = positionInfo.pendingFunding + positionInfo.owedRealizedPnL		
+    + owedRealizedPnLPositionUnit = totalFundingAndCarriedPnL / _setToken.totalSupply	
+  
+  + Trade
+
+    + *If long*: (when vBasePositionUnit is positive)
+
+    ```solidity	
+    openPositionParams = {
+      baseToken: vBase
+      isBaseToQuote: true, 
+      isExactInput: true, 
+      amount: vBaseTradeAmount
+      ...
+    }
+    
+    target = protocolAddresses.clearingHouse
+    deltaQuote = _setToken.invokeOpenPosition(target, openPositionParams) 
+    ```
+
+    + *If short*: ((when vBasePositionUnit is negative)
+
+    ```solidity	
+    openPositionParams = { 
+      baseToken: vBase
+      isBaseToQuote: false, 
+      isExactInput: false, 
+      amount: vBaseTradeAmount
+      ...
+    }
+    
+    target = protocolAddresses.clearingHouse
+    deltaQuote = _setToken.invokeOpenPosition(target, openPositionParams)
+    ```
+
+  + Calculate realized PnL and add to running total
+    + *If long*
+      + realizedPnL += reducedOpenNotional + deltaQuote
+
+    + *If short*
+      + realizedPnL += reducedOpenNotional - deltaQuote
+
++ Calculate amount of USDC to withdraw
+  + collateralPositionUnit = protocolAddresses.vault.balanceOf(_setToken) / _setToken.totalSupply
+  + usdcToWithdraw = 
+    + (collateralPositionUnit * _setTokenQuantity) + 
+    + (owedRealizedPnLPositionUnit * _setTokenQuantity) + 
+    + realizedPnL 
+
++ Set the external position unit for DIM 
+  + newUnit = usdcToWithdraw / _setTokenQuantity
+  + token = collateralToken[_setToken] 
+  + setToken.editExternalPositionUnit(token, address(this), newUnit)
+-----
+
+> **componentRedeemHook**: called during *DIM.resolveEquityPositions* via *DIM.externalPositionHook*. 
+
+```solidity
+function componentRedeemHook(
+  ISetToken _setToken, 
+  uint256 _setTokenQuantity, 
+  IERC20 _component
+)
+  external 
+  override 
+  onlyModule(_setToken)
+```
++ externalPositionUnit = _setToken.getExternalPositionRealUnit(_component, address(this))
++ usdcToWithdraw = externalPositionUnit * _setTokenQuantity
++ setToken.invokeWithdraw(usdcToWithdraw)
+-----
+
+> **deposit**: 
+
+```solidity
+function _deposit(
+  ISetToken _setToken, 
+  uint256 _collateralQuantityUnits
+) 
+  external
+  nonReentrant
+  onlyManagerAndValidSet(_setToken)
+```
+
++ notionalQuantity = _collateralQuantityUnits * _setToken.totalSupply
++ token = collateralToken[_setToken]
++ decimals = token.decimals()
++ notionalQuantityInCollateralDecimals = _formatCollateralToken(notionalQuantity, decimals)
++ _setToken.invokeApprove(collateralToken, protocolAddresses.vault, notionalQuantityInCollateralDecimals)
++ _setToken.invokeDeposit(protocolAddresses.vault, token, notionalQuantityInCollateralDecimals)
```

Comment:
Need to also update SetToken Position state here. Reduce Default USDC position and open/add to USDC external position